File: src/main/java/com/github/samarium150/structurescompass/util/sort/SourceCategory.java
package com.github.samarium150.structurescompass.util.sort;

import com.github.samarium150.structurescompass.util.StructureUtils;
import net.minecraft.client.resources.I18n;
import net.minecraft.world.gen.feature.structure.Structure;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

import javax.annotation.Nonnull;

/**
 * Class for sorting structures by their source mod
 */
@OnlyIn(Dist.CLIENT)
public final class SourceCategory implements Category {
    
    @Override
    public int compare(Structure<?> s1, Structure<?> s2) {
        return StructureUtils.getStructureSource(s1).compareTo(StructureUtils.getStructureSource(s2));
    }
    
    @Nonnull
    @Override
    public Category next() {
        return new DimensionCategory();
    }
    
    @Nonnull
    @Override
    public String getLocalizedName() {
        return I18n.get("string.structurescompass.source");
    }
}

File: src/main/kotlin/io/github/samarium150/minecraft/mod/structures_compass/client/util/sort/SourceCategory.kt
/*
 * Copyright (c) 2022 Samarium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
 */
package io.github.samarium150.minecraft.mod.structures_compass.client.util.sort

import io.github.samarium150.minecraft.mod.structures_compass.util.getNamespace
import io.github.samarium150.minecraft.mod.structures_compass.util.prefix
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.resource.language.I18n
import net.minecraft.world.gen.feature.StructureFeature

@Environment(EnvType.CLIENT)
object SourceCategory : Category {

    override val localizedName: String
        get() = I18n.translate("${prefix}source")

    override fun compare(s1: StructureFeature<*>, s2: StructureFeature<*>): Int {
        return s2.getNamespace()?.let { s1.getNamespace()?.compareTo(it) } ?: 0
    }

    override fun next(): Category {
        return DimensionCategory
    }

}
